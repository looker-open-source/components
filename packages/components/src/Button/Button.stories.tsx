/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { AddCircle, Delete } from '@styled-icons/material'
import type { Page } from 'puppeteer'
import React from 'react'
import type { Story } from '@storybook/react/types-6-0'
import { defaultArgTypes as argTypes } from '../../../../apps/storybook/src/defaultArgTypes'
import { Button } from './Button'
import type { ButtonProps } from './types'

const Template: Story<ButtonProps> = args => <Button {...args} />

export const Basic = Template.bind({})
Basic.args = {
  children: 'Button Text',
}

export const Focused = Template.bind({})
Focused.args = {
  ...Basic.args,
  margin: 'small',
}

const beforeScreenshot = async (page: Page) => {
  const button = await page.$('button')
  await button?.type(' ')
  await page.waitForTimeout(50)
}
Focused.parameters = {
  beforeScreenshot,
  docs: { disable: true },
}

export const Critical = Template.bind({})
Critical.args = {
  ...Basic.args,
  color: 'critical',
}

export const CriticalFocused = Template.bind({})
CriticalFocused.args = {
  ...Critical.args,
  margin: 'small',
}

CriticalFocused.parameters = {
  beforeScreenshot,
}

export const Neutral = Template.bind({})
Neutral.args = {
  ...Basic.args,
  color: 'neutral',
}

export const NeutralFocused = Template.bind({})
NeutralFocused.args = {
  ...Neutral.args,
  margin: 'small',
}

NeutralFocused.parameters = {
  beforeScreenshot,
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Basic.args,
  disabled: true,
}

export const IconBefore = Template.bind({})
IconBefore.args = {
  ...Basic.args,
  iconBefore: <AddCircle />,
}

export const IconAfter = Template.bind({})
IconAfter.args = {
  ...Basic.args,
  iconAfter: <Delete />,
}

export const FullWidth = Template.bind({})
FullWidth.args = {
  ...Basic.args,
  fullWidth: true,
}

export const XSmall = Template.bind({})
XSmall.args = {
  ...Basic.args,
  size: 'xsmall',
}

export const Small = Template.bind({})
Small.args = {
  ...Basic.args,
  size: 'small',
}

export const Large = Template.bind({})
Large.args = {
  ...Basic.args,
  size: 'large',
}

export default {
  argTypes,
  component: Button,
  title: 'Button',
}
