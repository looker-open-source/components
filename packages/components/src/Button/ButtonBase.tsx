/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  CompatibleHTMLProps,
  radii,
  RadiusSizes,
  reset,
  SemanticColors,
  space,
  SpaceProps,
} from '@looker/design-tokens'
import { rgba } from 'polished'
import React, { forwardRef, Ref, useState } from 'react'
import styled, { css } from 'styled-components'
import {
  minWidth,
  MinWidthProps,
  maxWidth,
  MaxWidthProps,
  WidthProps,
  width,
} from 'styled-system'
import { buttonSize, ButtonSizes, ButtonSizeProps } from './size'
import { ButtonIcon, buttonIcon, ButtonIconProps } from './icon'

export interface CustomizableButtonAttributes {
  borderRadius: RadiusSizes
}

export const CustomizableButtonAttributes: CustomizableButtonAttributes = {
  borderRadius: 'medium',
}

type ButtonColors = keyof SemanticColors

export interface ButtonBaseProps
  extends Omit<CompatibleHTMLProps<HTMLButtonElement>, 'type'>,
    ButtonSizeProps,
    MaxWidthProps,
    MinWidthProps,
    WidthProps,
    SpaceProps {
  type?: 'button' | 'submit' | 'reset'

  /**
   * Used along with `href` for an anchor link styled as a Button
   */
  forwardedAs?: 'a'

  /**
   * Defines the color of the button. Can be the string name of a color listed in the color theme, or a color object.
   * @default "primary"
   */
  color?: ButtonColors

  focusVisible?: boolean
}

export interface ButtonProps extends ButtonBaseProps, ButtonIconProps {
  size?: ButtonSizes
  /**
   * If true, the button's width will be set to 100%.
   */
  fullWidth?: boolean
}

export const buttonCSS = css<ButtonBaseProps>`
  ${reset}
  ${maxWidth}
  ${minWidth}
  ${width}

  ${props =>
    props.focusVisible &&
    `
    box-shadow: 0 0 0 0.15rem
      ${rgba(
        props.theme.colors.semanticColors[props.color || 'primary'].main,
        0.25
      )};
  `}

  align-items: center;
  border-radius: ${radii[CustomizableButtonAttributes.borderRadius]};
  cursor: pointer;
  display: inline-flex;
  font-weight: 600;
  justify-content: center;
  line-height: 1;
  outline: none;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;

  &[disabled] {
    cursor: default;
    filter: grayscale(0.3);
    opacity: 0.25;
  }

  ${buttonSize}
  ${space}
`

const ButtonOuter = styled.button<ButtonProps>`
  ${buttonCSS}
  ${props => props.fullWidth && `width: 100%;`}
`

const ButtonJSX = forwardRef(
  (props: ButtonProps, ref: Ref<HTMLButtonElement>) => {
    const {
      children,
      iconBefore,
      iconAfter,
      onBlur,
      onKeyDown,
      ...restProps
    } = props

    const [isFocusVisible, setFocusVisible] = useState(false)

    const handleOnKeyUp = (event: React.KeyboardEvent<HTMLButtonElement>) => {
      setFocusVisible(true)
      onKeyDown && onKeyDown(event)
    }

    const handleOnBlur = (event: React.FocusEvent<HTMLButtonElement>) => {
      setFocusVisible(false)
      onBlur && onBlur(event)
    }

    return (
      <ButtonOuter
        focusVisible={isFocusVisible}
        onKeyUp={handleOnKeyUp}
        onBlur={handleOnBlur}
        {...restProps}
        ref={ref}
      >
        {iconBefore && <ButtonIcon name={iconBefore} />}
        {children}
        {iconAfter && <ButtonIcon name={iconAfter} />}
      </ButtonOuter>
    )
  }
)

ButtonJSX.displayName = 'ButtonJSX'

export const ButtonBase = styled(ButtonJSX)<ButtonProps>`
  ${buttonIcon}
`

ButtonBase.defaultProps = { size: 'medium' }
