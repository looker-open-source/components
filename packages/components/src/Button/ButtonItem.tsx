/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { forwardRef, MouseEvent, Ref, useContext, useState } from 'react'
import styled from 'styled-components'
import {
  CompatibleHTMLProps,
  space,
  SpaceProps,
  omitStyledProps,
} from '@looker/design-tokens'
import { inputHeight } from '../Form/Inputs/height'
import { ButtonSetContext } from './ButtonSetContext'

export interface ButtonItemProps
  extends SpaceProps,
    Omit<CompatibleHTMLProps<HTMLButtonElement>, 'type' | 'aria-pressed'> {
  value?: string
  focusVisible?: boolean
}

const ButtonLayout = forwardRef(
  (
    { children, onClick, value, onBlur, onKeyUp, ...props }: ButtonItemProps,
    ref: Ref<HTMLButtonElement>
  ) => {
    const { disabled, value: contextValue, onItemClick } = useContext(
      ButtonSetContext
    )

    const [isFocusVisible, setFocusVisible] = useState(false)

    const handleOnKeyUp = (event: React.KeyboardEvent<HTMLButtonElement>) => {
      setFocusVisible(true)
      onKeyUp && onKeyUp(event)
    }

    const handleOnBlur = (event: React.FocusEvent<HTMLButtonElement>) => {
      setFocusVisible(false)
      onBlur && onBlur(event)
    }

    function handleClick(e: MouseEvent<HTMLButtonElement>) {
      onClick && onClick(e)
      if (!e.defaultPrevented) {
        onItemClick && onItemClick(e)
      }
    }

    const itemValue =
      value !== undefined ? value : typeof children === 'string' ? children : ''

    const selected = contextValue
      ? typeof contextValue === 'string'
        ? contextValue === itemValue
        : contextValue.includes(itemValue)
      : false

    return (
      <ButtonOuter
        aria-pressed={selected}
        ref={ref}
        onClick={handleClick}
        value={itemValue}
        disabled={disabled}
        focusVisible={isFocusVisible}
        onKeyUp={handleOnKeyUp}
        onBlur={handleOnBlur}
        {...omitStyledProps(props)}
      >
        {children}
      </ButtonOuter>
    )
  }
)

ButtonLayout.displayName = 'ButtonLayout'

const ButtonOuter = styled.button<ButtonItemProps>`
  ${(props) =>
    props.focusVisible &&
    `box-shadow: 0 0 0.5px 1px ${props.theme.colors.keyFocus}`}
`

export const ButtonItem = styled(ButtonLayout)`
  align-items: center;
  background: transparent;
  border: none;
  color: ${({ theme }) => theme.colors.text3};
  cursor: pointer;
  display: inline-flex;
  font-family: ${({ theme }) => theme.fonts.brand};
  font-size: ${({ theme }) => theme.fontSizes.small};
  height: ${inputHeight};
  justify-content: center;
  margin: 0;
  padding: 0 ${({ theme }) => theme.space.small};
  transition: background ${({ theme }) => theme.transitions.quick}ms ease;
  user-select: none;

  ${space}

  &:active,
  &:focus,
  &:hover {
    background: ${({ theme }) => theme.colors.neutralSubtle};
    color: ${({ theme }) => theme.colors.text5};
  }

  &:focus {
    outline: none;
  }

  &[disabled] {
    color: ${(props) => props.theme.colors.text1};

    &:hover {
      background: inherit;
    }
    cursor: default;
  }

  &[aria-pressed='true'] {
    background: ${({ theme }) => theme.colors.keySubtle};
    color: ${({ theme }) => theme.colors.key};

    &[disabled] {
      background: ${({ theme }) => theme.colors.keySubtle};
      color: ${({ theme }) => theme.colors.keyFocus};
    }
  }
`
