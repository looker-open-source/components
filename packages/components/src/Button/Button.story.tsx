/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { Story } from '@storybook/react/types-6-0'
import { Button } from './Button'
import { ButtonProps } from './types'

const Template: Story<ButtonProps> = (args) => <Button {...args} />

export const Basic = Template.bind({})
Basic.args = {
  children: 'Button Text',
}

export const Critical = Template.bind({})
Critical.args = {
  ...Basic.args,
  color: 'critical',
}

export const Neutral = Template.bind({})
Neutral.args = {
  ...Basic.args,
  color: 'neutral',
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Basic.args,
  disabled: true,
}

export const IconBefore = Template.bind({})
IconBefore.args = {
  ...Basic.args,
  iconBefore: 'CircleAdd',
}

export const IconAfter = Template.bind({})
IconAfter.args = {
  ...Basic.args,
  iconAfter: 'Trash',
}

export const FullWidth = Template.bind({})
FullWidth.args = {
  ...Basic.args,
  fullWidth: true,
}

export const XSmall = Template.bind({})
XSmall.args = {
  ...Basic.args,
  size: 'xsmall',
}

export const Small = Template.bind({})
Small.args = {
  ...Basic.args,
  size: 'small',
}

export const Large = Template.bind({})
Large.args = {
  ...Basic.args,
  size: 'large',
}

export default {
  component: Button,
  title: 'Button',
}
