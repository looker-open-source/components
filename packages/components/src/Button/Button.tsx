/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import { ButtonBase } from './ButtonBase'

export const Button = styled(ButtonBase)`
  background: ${({ theme, color = 'key' }) => theme.colors[color]};
  border: 1px solid ${({ theme, color = 'key' }) => theme.colors[color]};
  color: ${({ theme, color = 'key' }) => theme.colors[`${color}Text`]};

  &:hover,
  &:focus,
  &.hover {
    background: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Interactive`]};
    border-color: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Interactive`]};
  }

  &[aria-expanded='true'],
  &:active,
  &.active {
    background: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Pressed`]};
    border-color: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Pressed`]};
  }

  &[disabled] {
    &:hover,
    &:active,
    &:focus {
      background-color: ${({ theme, color = 'key' }) => theme.colors[color]};
      border-color: ${({ theme }) => theme.colors.ui3};
    }
  }
`
