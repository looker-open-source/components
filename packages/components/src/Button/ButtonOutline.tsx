/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import { ButtonBase } from './ButtonBase'

export const ButtonOutline = styled(ButtonBase)`
  background: ${({ theme, color = 'primary' }) =>
    theme.colors.semanticColors[color].white};
  border: 1px solid
    ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].borderColor};
  color: ${({ theme, color = 'primary' }) =>
    theme.colors.semanticColors[color].main};

  &:hover,
  &:focus,
  &.hover {
    background: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].white};
    border-color: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].main};
    color: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].darker};
  }

  &:active,
  &.active {
    background: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].main};
    border-color: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].main};
    color: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].text};
  }

  &[disabled] {
    &:hover,
    &:active,
    &:focus {
      background-color: ${({ theme, color = 'primary' }) =>
        theme.colors.semanticColors[color].white};
      border-color: ${({ theme, color = 'primary' }) =>
        theme.colors.semanticColors[color].borderColor};
      color: ${({ theme, color = 'primary' }) =>
        theme.colors.semanticColors[color].main};
    }
  }
`
