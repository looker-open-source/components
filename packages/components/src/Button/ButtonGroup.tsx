/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import uniqueId from 'lodash/uniqueId'
import xor from 'lodash/xor'
import React, { ChangeEvent, FC, forwardRef, Ref } from 'react'
import styled from 'styled-components'
import { useControlWarn } from '../utils'
import { ButtonItem, ButtonItemLabel } from './ButtonItem'
import { ButtonGroupOrToggleProps, ButtonSet } from './ButtonSet'

const ButtonGroupFactory = forwardRef(
  (
    { onChange, value, ...props }: ButtonGroupOrToggleProps<string[]>,
    ref: Ref<HTMLDivElement>
  ) => {
    const isControlled = useControlWarn({
      controllingProps: ['onChange', 'value'],
      isControlledCheck: () => onChange !== undefined,
      name: 'ButtonGroup',
    })

    function handleChange(e: ChangeEvent<HTMLInputElement>) {
      const newValue = xor(value, [e.target.value])
      onChange && onChange(newValue)
    }

    return (
      <ButtonSet
        {...props}
        ref={ref}
        {...(isControlled
          ? {
              onChange: handleChange,
              value,
            }
          : { name: props.name || uniqueId() })}
      />
    )
  }
)

export const ButtonGroup = styled<FC<ButtonGroupOrToggleProps<string[]>>>(
  ButtonGroupFactory
)`
  ${ButtonItem} {
    height: 36px;
    border-radius: 4px;

    + ${ButtonItem} {
      margin-left: ${(props) => props.theme.space.xxsmall};
    }
  }

  ${ButtonItemLabel} {
    border-style: solid;
    border-width: 1px;
  }
`
