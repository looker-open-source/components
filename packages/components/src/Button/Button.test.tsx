/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { assertSnapshotShallow } from '@looker/components-test-utils'
import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { theme } from '@looker/design-tokens'
import { Button } from './Button'

test('Button is rendered ', () => {
  assertSnapshotShallow(<Button>click here</Button>)
})

test('Button works with color danger', () => {
  const { getByText } = render(
    <ThemeProvider theme={theme}>
      <Button color="danger">danger</Button>
    </ThemeProvider>
  )

  expect(getByText('danger')).toMatchSnapshot()
})

test('Button disable', () => {
  const { getByText } = render(
    <ThemeProvider theme={theme}>
      <Button disabled>disabled</Button>
    </ThemeProvider>
  )

  expect(getByText('disabled')).toMatchSnapshot()
})

test('Button accepts a className prop', () => {
  const { container } = render(
    <ThemeProvider theme={theme}>
      <Button className="foo">button with class</Button>
    </ThemeProvider>
  )

  expect(container.firstChild).toHaveClass('foo')
})

test('Button validates all sizes', () => {
  const { getByText } = render(
    <ThemeProvider theme={theme}>
      <>
        <Button size={'xsmall'}>xsmall button</Button>
        <Button size={'small'}>small button</Button>
        <Button size={'medium'}>medium button</Button>
        <Button size={'large'}>large button</Button>
      </>
    </ThemeProvider>
  )

  expect(getByText('xsmall button')).toMatchSnapshot()
  expect(getByText('small button')).toMatchSnapshot()
  expect(getByText('medium button')).toMatchSnapshot()
  expect(getByText('large button')).toMatchSnapshot()
})

test('Button can be full width', () => {
  const { getByText } = render(
    <ThemeProvider theme={theme}>
      <Button fullWidth>Full Width</Button>
    </ThemeProvider>
  )

  expect(getByText('Full Width')).toMatchSnapshot()
})

test('Button Focus: renders outline when tabbing into focus, but not when clicking', () => {
  const { getByText } = render(
    <ThemeProvider theme={theme}>
      <>
        <Button>button</Button>
        <Button>focus</Button>
      </>
    </ThemeProvider>
  )

  fireEvent.click(getByText('button'))
  expect(getByText('focus')).toMatchSnapshot()

  fireEvent.keyUp(getByText('focus'), { charCode: 9, code: 9, key: 'Tab' })
  expect(getByText('focus')).toMatchSnapshot()
})
