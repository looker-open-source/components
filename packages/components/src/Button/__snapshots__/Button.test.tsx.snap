// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Focus: renders outline when tabbing into focus, but not when clicking 1`] = `
.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.25rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: Roboto,'Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-weight: 500;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  outline: none;
  -webkit-transition: border 80ms;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;
  font-size: 0.875rem;
  height: 36px;
  padding-left: 24px;
  padding-right: 24px;
}

.c0[disabled] {
  cursor: default;
  -webkit-filter: grayscale(0.3);
  filter: grayscale(0.3);
  opacity: 0.25;
}

.c1 {
  background: #6C43E0;
  border: 1px solid #6C43E0;
  color: #FFFFFF;
}

.c1:hover,
.c1:focus,
.c1.hover {
  background: #7a55e3;
  border-color: #7a55e3;
}

.c1[aria-expanded='true'],
.c1:active,
.c1.active {
  background: #5424db;
  border-color: #5424db;
}

.c1[disabled]:hover,
.c1[disabled]:active,
.c1[disabled]:focus {
  background-color: #6C43E0;
  border-color: #C1C6CC;
}

<button
  class="c0 c1"
>
  focus
</button>
`;

exports[`Button Focus: renders outline when tabbing into focus, but not when clicking 2`] = `
<button
  class="ButtonBase__ButtonOuter-sc-1bpio6j-0 kFFdak ButtonBase-sc-1bpio6j-1 Button-sc-18euc9m-0 dXZmpP"
>
  focus
</button>
`;
